std/digest/xxh.d

```diff
@@ -0,0 +1,3177 @@
+enum XXH_PRIME64_2 = 0xC2B2AE3D27D4EB4F; /** 0b1100001010110010101011100011110100100111110101001110101101001111 */
+enum XXH_PRIME64_3 = 0x165667B19E3779F9; /** 0b0001011001010110011001111011000110011110001101110111100111111001 */
+enum XXH_PRIME64_4 = 0x85EBCA77C2B2AE63; /** 0b1000010111101011110010100111011111000010101100101010111001100011 */
+enum XXH_PRIME64_5 = 0x27D4EB2F165667C5; /** 0b0010011111010100111010110010111100010110010101100110011111000101 */
+
+private ulong XXH64_round(ulong acc, ulong input) @safe pure nothrow @nogc
+{
+    acc += input * XXH_PRIME64_2;
+    acc = rol(acc, 31);
+    acc *= XXH_PRIME64_1;
+    return acc;
+}
+
+private ulong XXH64_mergeRound(ulong acc, ulong val) @safe pure nothrow @nogc
+{
+    val = XXH64_round(0, val);
+    acc ^= val;
+    acc = acc * XXH_PRIME64_1 + XXH_PRIME64_4;
+    return acc;
+}
+
+private ulong XXH64_avalanche(ulong hash) @safe pure nothrow @nogc
+{
+    hash ^= hash >> 33;
+    hash *= XXH_PRIME64_2;
+    hash ^= hash >> 29;
+    hash *= XXH_PRIME64_3;
+    hash ^= hash >> 32;
+    return hash;
+}
+
+ulong XXH_get64bits(const void* p, XXH_alignment align_) @safe pure nothrow @nogc
+{
+    return XXH_readLE64_align(p, align_);
+}
+
+/** Processes the last 0-31 bytes of @p ptr.
+ *
+ * There may be up to 31 bytes remaining to consume from the input.
+ * This final stage will digest them to ensure that all input bytes are present
+ * in the final mix.
+ *
+ * Param: hash The hash to finalize.
+ * Param: ptr The pointer to the remaining input.
+ * Param: len The remaining length, modulo 32.
+ * Param: align Whether @p ptr is aligned.
+ * Return: The finalized hash
+ * See: XXH32_finalize().
+ */
+private ulong XXH64_finalize(ulong hash, const(ubyte)* ptr, size_t len, XXH_alignment align_)
+    @trusted pure nothrow @nogc
+{
+    if (ptr == null)
+        assert(len == 0, "input null ptr only allowed with len == 0");
+
+    len &= 31;
+    while (len >= 8)
+    {
+        ulong k1 = XXH64_round(0, XXH_get64bits(ptr, align_));
+        ptr += 8;
+        hash ^= k1;
+        hash = rol(hash, 27) * XXH_PRIME64_1 + XXH_PRIME64_4;
+        len -= 8;
+    }
+    if (len >= 4)
+    {
+        hash ^= cast(ulong)(XXH_get32bits(ptr, align_)) * XXH_PRIME64_1;
+        ptr += 4;
+        hash = rol(hash, 23) * XXH_PRIME64_2 + XXH_PRIME64_3;
+        len -= 4;
+    }
+    while (len > 0)
+    {
+        hash ^= (*ptr++) * XXH_PRIME64_5;
+        hash = rol(hash, 11) * XXH_PRIME64_1;
+        --len;
+    }
+    return XXH64_avalanche(hash);
+}
+
+/** The implementation for XXH64().
+ *
+ * Param: input , len , seed Directly passed from XXH64().
+ * Param: align Whether @p input is aligned.
+ * Return: The calculated hash.
+ */
+pragma(inline, true)
+private ulong XXH64_endian_align(const(ubyte)* input, size_t len,
+        ulong seed, XXH_alignment align_) @trusted pure nothrow @nogc
+{
+    ulong h64;
+    if (input == null)
+        assert(len == 0, "input null ptr only allowed with len == 0");
```

Comment:
Redundant if statement. I think this is equivelent, if I'm not missing something:

```suggestion
    assert(input !is null && !len, "input null ptr only allowed with len == 0");
```